### ISSUE TITLE
Add @Nullable/@NotNull to API (for Kotlin, etc)

### ISSUE THREAD

**Issue by geekley** [jhy/jsoup#1264]
Would you consider adding [nullability annotations]( https://kotlinlang.org/docs/reference/java-interop.html#nullability-annotations ) to the API? The official page says that jsoup is compatible with Kotlin, so it would be great to have its null-safety mechanisms when using the library.

**Comment by HansBrende** [comment]
+1; this would helpful even when writing vanilla Java (accurate IntelliJ warnings, etc.)

**Comment by jhy** [comment]
Sure, I like the idea. I guess we want to implement like in https://github.com/Kotlin/KEEP/blob/41091f1cc7045142181d8c89645059f4a15cc91a/proposals/jsr-305-custom-nullability-qualifiers.md ?

On first read through I didn't quite follow the proposal for migrations. Can you suggest some libraries that have done a good job of adding these annotations and releases?

**Comment by geekley** [comment]
Well, on Android Studio (and IntelliJ?) it should be easy to infer these annotations, at least the editor seems to infer some when browsing an external library, so I assume it probably has some functionality to make it easier to add them (probably the JetBrains annotations).

**Comment by jhy** [comment]
Yep got that, but looking specifically for other libraries that have added the annotations to help Kotlin and how they managed _migrations_ through releases.

**Comment by fvasco** [comment]
Hi, @jhy,
it is possible to use plain JSR-305 annotations, you are not forced to wrote your own. These works well with both Java and Kotlin.
Moreover, please consider that adding/removing annotation doesn't break binary compatibility (nor Java source compatibility).
Instead, it is possible to break Kotlin source compatibility, in case of function invocation on a nullable reference, so a developer can get an error at compile time.

Personally I discourage any automatic inference, this is a good moment to spot hidden bugs. Instead it may be enough to annotate only public interfaces.

Regarding your question, take a look here:
https://github.com/spring-projects/spring-framework/issues/20099

**Comment by geekley** [comment]
> Instead it may be enough to annotate only public interfaces.

Also, an alternative way to provide annotations is simply just writing the interfaces in Kotlin. That is, rewrite any existing interfaces in Kotlin. For other classes, you can write new interfaces in Kotlin and make the classes implement them.
It's totally possible to mix Kotlin and Java code.
It might require some rewriting but Kotlin syntax is a lot cleaner than annotations everywhere and you can still do most of it gradually. And if you ever want to also convert implementation code to Kotlin to benefit from null-safety checks (and catch potentially hidden bugs) you would have already done part of the work.

**Comment by fvasco** [comment]
I disagree, @geekley.
I don't see a valid reason for introduce a new technology due this task.
Moreover, rewriting interface in Kotlin may lead to unexpected results, i.e.: Kotlin cannot return `void` in its interfaces and this breaks binary compatibility, so this is a no-go for me.

JSR-305 is well supported on both Java and Kotlin, it does not require a big effort and it is pretty safe to introduce, it is enough.

**Comment by geekley** [comment]
Oh, that's true. I didn't think about that.

**Comment by jhy** [comment]
Thanks for the notes. Will plan to get started on this in the next (1.14.1) release, and will appreciate non-automated PRs. Piecemeal is fine. Very much agree with your line of thinking, @fvasco

**Comment by jhy** [comment]
I added a first pass at this with #1467. Only covers Jsoup.class now, but I / we can easily add coverage over time.

I did start the PR with including Kotlin tests, but didn't find it added much value after confirming that the nonnull assertions changed the returns from Document! to Document. I was a bit surprised by the Kotlin-lang choice of not warning on unsafe use of ! platform types, and the apparent inability to set IDE or compiler warnings for that.
